### ISSUE TITLE
Bug: [@typescript-eslint/no-unnecessary-condition] False positive when enable tsconfig noUncheckedIndexedAccess and branded string key

### ISSUE THREAD

**Issue by yf-yang** [typescript-eslint/typescript-eslint#7700]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.0.2&fileType=.tsx&code=C4TwDgpgBAQgTgQwHYBMIoNIRFAvFAZ2DgEskBzKAMigG8oB9AI0VQC4oBXVCAMzPRQAvgG4AUGLLAIcXggDG0ACoQAtmADyTAFYR5wOmKhQA9p2lwA-BwBKek3BQAeeMjSZsAGigBJJEhktXX0APnEhCSkZOUVff0CdPQNaIygyALgOVx4PEHCJeRMkIigEDhV1IKS8OlExQuKDJhqAcgITVWgAa2wW0oJYVncsPPqikvkahAA6MwtLaYBtJgBdBfSZESA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2uGw5E0aB2iRVEALoBfEIaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false

### Repro Code

```typescript
type BrandedKey = string & { _brand: undefined };

interface TempObject {
  outer?: Record<BrandedKey, InnerObject>;
}

interface InnerObject {
  inner: BrandedKey;
}

const a: TempObject = {};
const b = 'some key' as BrandedKey;
const c = a.outer?.[b]?.inner;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": [
      "error"
    ]}
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "noUncheckedIndexedAccess": true
  }
}
```

### Expected Result

The line
``` typescript
const c = a.outer?.[b]?.inner;
```
does not have any error

### Actual Result

The `?.` operator before `.inner` is reported as an error (false positive)
``` typescript
const c = a.outer?.[b]?.inner;
```

**Comment by yf-yang** [comment]
I got the root cause
https://github.com/typescript-eslint/typescript-eslint/blob/4a6d841f6def9944efc98016ec4c5740fa019f64/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L536-L542
Here `typeName` is `string` and `checker.getIndexInfoOfType` cannot get the correct info, because TypeScript compares the types by deep equality, so it cannot find the correct info given string type.

``` typescript
// the keyType is string type here
function getIndexInfoOfType(type: Type, keyType: Type): IndexInfo | undefined {
    return findIndexInfo(getIndexInfosOfType(type), keyType);
}

function findIndexInfo(indexInfos: readonly IndexInfo[], keyType: Type) {
    return find(indexInfos, info => info.keyType === keyType);
}
```

However, I tried other variants of `getIndexInfoOfType`, they are still not working because `info.keyType` is `string & { _brand: undefined; }` and `propertyType` at line 536 of typescript-eslint code is `BrandedKey` and they have different type ID, so I suspect `getIndexInfoOfType` is not the right API to call, not sure how TypeScript implements that part.

Do you have any suggestion?

**Comment by yf-yang** [comment]
``` typescript
const c: Record<BrandedKey, InnerObject> = {};
c[b]?.inner;
```

This case is fine because typescript-eslint can tell `c[b]` is optional.
The reported case is much harder because it is in an optional chain, so typescript-eslint still get an `InnerObject | undefined` type signature, but it has to distinguish if the `undefined` is inherited from previous optional code, or originates from the new member access. Maybe I need to check TypeScript's implementation for such case.

**Comment by QuentinLemCode** [comment]
Hello
I have a similar case, found by a co-worker.
Here is the playground : https://typescript-eslint.io/play/#ts=5.2.2&showAST=ts&fileType=.tsx&code=C4TwDgpgBAYghgYwsKBeKBvAUFKBrASwDsATAfgC4oBnYAJ2IHMsBfLLUSKAcToHsArmDSYcUAGaJk1KvCTAA2gF1W7BHyK0oAWxC9BYGT35DlIhWIwSpwIwquFSVAOTA%2Bb51BYAaTCyVe3pbW8nYOxCQuwATRnj5%2BAWwqWOqaKOJ8fCK6%2BkLUYgB04gA2cMAAsnBgABTVMRDaAJRoAHxQ9dpFNtSNhXQQJAJItf3UAsXAvggCdP1EwABqcMUCEM2obdi4uATiUNXTsxDzSysQBY4kzdhi21AA9PdQABYQ-VCCdFDqJNAZdEhqDpoG4oHASCQoMBXlAiHwALQCIhECCA6hwOggeGpEgxAgaW7baqjcaKQ5zRbLVYXCIBMhkdDKRoFMACajPA4zCmnVa9O5sQn9YAzIhQEkTMTxDAsMFAgBKqL4dBIAB5aAwiIxfHJkMoWo0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false

Can you take a look please and tell me if you think it's the same issue ?

**Comment by yf-yang** [comment]
> Hello I have a similar case, found by a co-worker. Here is the playground : https://typescript-eslint.io/play/#ts=5.2.2&showAST=ts&fileType=.tsx&code=C4TwDgpgBAYghgYwsKBeKBvAUFKBrASwDsATAfgC4oBnYAJ2IHMsBfLLUSKAcToHsArmDSYcUAGaJk1KvCTAA2gF1W7BHyK0oAWxC9BYGT35DlIhWIwSpwIwquFSVAOTA%2Bb51BYAaTCyVe3pbW8nYOxCQuwATRnj5%2BAWwqWOqaKOJ8fCK6%2BkLUYgB04gA2cMAAsnBgABTVMRDaAJRoAHxQ9dpFNtSNhXQQJAJItf3UAsXAvggCdP1EwABqcMUCEM2obdi4uATiUNXTsxDzSysQBY4kzdhi21AA9PdQABYQ-VCCdFDqJNAZdEhqDpoG4oHASCQoMBXlAiHwALQCIhECCA6hwOggeGpEgxAgaW7baqjcaKQ5zRbLVYXCIBMhkdDKRoFMACajPA4zCmnVa9O5sQn9YAzIhQEkTMTxDAsMFAgBKqL4dBIAB5aAwiIxfHJkMoWo0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false
> 
> Can you take a look please and tell me if you think it's the same issue ?

I don't think so, actually this issue is related to issue https://github.com/typescript-eslint/typescript-eslint/issues/2247, which is an issue related to the `checkOptionalMemberExpression`. According to the estree viewer, it seems your issue is related to `checkAssignmentExpression`.

https://github.com/typescript-eslint/typescript-eslint/blob/d948dc4a21ad8e15eec152c0cf2fdda819ea4a3a/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L666-L678 

However, the solution may be related to #2247's, I am not sure. I'm still on holiday, so if you can wait two more days (and no one has figured it out yet), I can take a look of the actual problem :)

**Comment by QuentinLemCode** [comment]
> > Hello I have a similar case, found by a co-worker. Here is the playground : https://typescript-eslint.io/play/#ts=5.2.2&showAST=ts&fileType=.tsx&code=C4TwDgpgBAYghgYwsKBeKBvAUFKBrASwDsATAfgC4oBnYAJ2IHMsBfLLUSKAcToHsArmDSYcUAGaJk1KvCTAA2gF1W7BHyK0oAWxC9BYGT35DlIhWIwSpwIwquFSVAOTA%2Bb51BYAaTCyVe3pbW8nYOxCQuwATRnj5%2BAWwqWOqaKOJ8fCK6%2BkLUYgB04gA2cMAAsnBgABTVMRDaAJRoAHxQ9dpFNtSNhXQQJAJItf3UAsXAvggCdP1EwABqcMUCEM2obdi4uATiUNXTsxDzSysQBY4kzdhi21AA9PdQABYQ-VCCdFDqJNAZdEhqDpoG4oHASCQoMBXlAiHwALQCIhECCA6hwOggeGpEgxAgaW7baqjcaKQ5zRbLVYXCIBMhkdDKRoFMACajPA4zCmnVa9O5sQn9YAzIhQEkTMTxDAsMFAgBKqL4dBIAB5aAwiIxfHJkMoWo0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false
> > Can you take a look please and tell me if you think it's the same issue ?
> 
> I don't think so, actually this issue is related to issue #2247, which is an issue related to the `checkOptionalMemberExpression`. According to the estree viewer, it seems your issue is related to `checkAssignmentExpression`.
> 
> https://github.com/typescript-eslint/typescript-eslint/blob/d948dc4a21ad8e15eec152c0cf2fdda819ea4a3a/packages/eslint-plugin/src/rules/no-unnecessary-condition.ts#L666-L678
> 
> However, the solution may be related to #2247's, I am not sure. I'm still on holiday, so if you can wait two more days (and no one has figured it out yet), I can take a look of the actual problem :)

I can wait, we have found another way to handle that :) 

Thanks for taking a look ! :heart:

**Comment by Josh-Cena** [comment]
@QuentinLemCode: your issue is https://github.com/typescript-eslint/typescript-eslint/issues/6635. Just use a disable comment for the time being.

**Comment by QuentinLemCode** [comment]
Thanks I will follow this issue 👍
